src/Shared/TypeLoader.cs

```diff
@@ -169,8 +172,33 @@ private static Assembly LoadAssembly(AssemblyLoadInfo assemblyLoadInfo)
                 // NOTE: don't use ErrorUtilities.VerifyThrowFileExists() here because that will hit the disk again
                 throw new FileNotFoundException(null, assemblyLoadInfo.AssemblyLocation, e);
             }
+        }
 
-            return loadedAssembly;
+        private static Assembly LoadAssemblyUsingMetadataLoadContext(AssemblyLoadInfo assemblyLoadInfo)
+        {
+            string path = assemblyLoadInfo.AssemblyFile;
+
+            if (path is null)
+            {
+#if NETFRAMEWORK
+                AppDomainSetup setup = AppDomain.CurrentDomain.SetupInformation;
+                setup.LoaderOptimization = LoaderOptimization.SingleDomain;
+                AppDomain appDomain = AppDomain.CreateDomain("appDomainToFindPath", null, setup);
+                path = appDomain.Load(new AssemblyName(assemblyLoadInfo.AssemblyName)).Location;
+                AppDomain.Unload(appDomain);
+#else
+                AssemblyLoadContext alc = new("loadContextToFindPath", true);
+                path = alc.LoadFromAssemblyName(new AssemblyName(assemblyLoadInfo.AssemblyName)).Location;
+                alc.Unload();
+#endif
+            }
+
+            string[] runtimePaths = Directory.GetFiles(RuntimeEnvironment.GetRuntimeDirectory(), "*.dll");
+            List<string> localPaths = new(Directory.GetFiles(Path.GetDirectoryName(path), "*.dll"));
+            localPaths.AddRange(runtimePaths);
+
+            MetadataLoadContext loadContext = new(new PathAssemblyResolver(localPaths));
```

Comment:
I could at least move around the assembly after having used it, so I'd assume this works. I'm assuming it's automatically disposed of as soon as it goes out of scope.

Reply:
> I'm assuming it's automatically disposed of as soon as it goes out of scope.

This is not how disposal works in .NET. You must manually dispose objects that are `IDisposable` to control the time that unmanaged resources get freed.

Typically, the finalizer of an object will free resources at GC time, so that handles aren't leaked. That may be happening in your case, but we should not rely on it.

Reply:
I checked what would happen if I explicitly disposed of the MetadataLoadContext before returning from this method, and it failed with:
This object is no longer valid because the MetadataLoadContext that created it has been disposed.

Not a change I can make. My vague understanding is that MetadataLoadContext loads the assembly from a "copied" version of the assembly, in which case it never loads the "real" assembly, and it's fine for it to keep its lock on the fake one.